xtrack/beam_elements/elements_src/henonmap.h

```diff
@@ -0,0 +1,184 @@
+
+        double const multipole_scale = 1.0 / (1.0 + delta);
+        for (int n = 0; n < n_turns; n++)
+        {
+
+            double fx = 0;
+            for (int i = 0; i < n_fx_coeffs; i++)
+            {
+                double prod = fx_coeffs[i] * multipole_scale;
+                int x_power = fx_x_exps[i];
+                int y_power = fx_y_exps[i];
+                for (int j = 0; j < x_power; j++)
+                {
+                    prod *= (sqrt_beta_x * x_hat);
+                }
+                for (int j = 0; j < y_power; j++)
+                {
+                    prod *= (sqrt_beta_y * y_hat);
+                }
+                fx += prod;
+            }
+            double fy = 0;
+            for (int i = 0; i < n_fy_coeffs; i++)
+            {
+                double prod = fy_coeffs[i] * multipole_scale;
+                int x_power = fy_x_exps[i];
+                int y_power = fy_y_exps[i];
+                for (int j = 0; j < x_power; j++)
+                {
+                    prod *= (sqrt_beta_x * x_hat);
+                }
+                for (int j = 0; j < y_power; j++)
+                {
+                    prod *= (sqrt_beta_y * y_hat);
+                }
+                fy += prod;
+            }
+            fx *= sqrt_beta_x;
+            fy *= sqrt_beta_y;
+
+            double curr_cos_omega_x, curr_sin_omega_x, curr_cos_omega_y, curr_sin_omega_y;
+            if (domegax == 0)
+            {
+                curr_cos_omega_x = cos_omega_x;
+                curr_sin_omega_x = sin_omega_x;
+            }
+            else
+            {
+                double cos_domega_x = cos(domegax * delta);
+                double sin_domega_x = sin(domegax * delta);
+                curr_cos_omega_x = cos_omega_x * cos_domega_x - sin_omega_x * sin_domega_x;
+                curr_sin_omega_x = sin_omega_x * cos_domega_x + cos_omega_x * sin_domega_x;
+            }
+            if (domegay == 0)
+            {
+                curr_cos_omega_y = cos_omega_y;
+                curr_sin_omega_y = sin_omega_y;
+            }
+            else
+            {
+                double cos_domega_y = cos(domegay * delta);
+                double sin_domega_y = sin(domegay * delta);
+                curr_cos_omega_y = cos_omega_y * cos_domega_y - sin_omega_y * sin_domega_y;
+                curr_sin_omega_y = sin_omega_y * cos_domega_y + cos_omega_y * sin_domega_y;
+            }
+
+            double x_hat_new, px_hat_new, y_hat_new, py_hat_new;
+            x_hat_new = curr_cos_omega_x * (x_hat - x_hat_f) + curr_sin_omega_x * (px_hat - px_hat_f + fx) + x_hat_f;
+            px_hat_new = -curr_sin_omega_x * (x_hat - x_hat_f) + curr_cos_omega_x * (px_hat - px_hat_f + fx) + px_hat_f;
+            y_hat_new = curr_cos_omega_y * y_hat + curr_sin_omega_y * (py_hat + fy);
+            py_hat_new = -curr_sin_omega_y * y_hat + curr_cos_omega_y * (py_hat + fy);
```

Comment:
These could be `const`, but maybe they aren't needed? You could write directly to the target variables and put a comment like "update normalised coordinates".

Reply:
I made them `const` now, but they are needed, because e.g. I use the value of `x_hat` to calculate `px_hat_new` etc.